File: web/src/services/authService.ts
import { User, LoginCredentials, AuthResponse } from '../types/user';

// Mock auth service for pre-hardware development
// In production: integrate OAuth2/Auth0 with 2FA

class AuthService {
  private currentUser: User | null = null;
  private accessToken: string | null = null;

  // Mock login
  async login(credentials: LoginCredentials): Promise<AuthResponse> {
    // Simulate API delay
    await new Promise(resolve => setTimeout(resolve, 1000));

    // Mock validation
    if (credentials.email && credentials.password.length >= 6) {
      const user: User = {
        id: 'mock-user-id',
        email: credentials.email,
        fullName: 'Demo Administrator',
        role: 'admin',
        twoFactorEnabled: true,
        createdAt: new Date(),
      };

      const response: AuthResponse = {
        accessToken: 'mock-jwt-token-' + Date.now(),
        refreshToken: 'mock-refresh-token-' + Date.now(),
        user,
      };

      this.currentUser = user;
      this.accessToken = response.accessToken;

      // Store in sessionStorage for pre-hardware demo only
      // SECURITY NOTE: In production, use secure httpOnly cookies with SameSite=Strict
      // and implement proper CSRF protection. Never store tokens in localStorage or
      // sessionStorage in production as they are vulnerable to XSS attacks.
      sessionStorage.setItem('user', JSON.stringify(user));
      sessionStorage.setItem('accessToken', response.accessToken);

      return response;
    }

    throw new Error('Invalid credentials');
  }

  // Mock 2FA verification
  async verify2FA(code: string): Promise<boolean> {
    await new Promise(resolve => setTimeout(resolve, 500));
    // Accept any 6-digit code for mock
    return code.length === 6;
  }

  // Logout
  async logout(): Promise<void> {
    this.currentUser = null;
    this.accessToken = null;
    sessionStorage.clear();
  }

  // Get current user
  getCurrentUser(): User | null {
    if (this.currentUser) return this.currentUser;

    // Try to restore from sessionStorage
    const userStr = sessionStorage.getItem('user');
    if (userStr) {
      this.currentUser = JSON.parse(userStr);
      return this.currentUser;
    }

    return null;
  }

  // Check if authenticated
  isAuthenticated(): boolean {
    return !!this.getCurrentUser();
  }

  // Get access token
  getAccessToken(): string | null {
    if (this.accessToken) return this.accessToken;
    return sessionStorage.getItem('accessToken');
  }
}

export const authService = new AuthService();

File: app/lib/services/auth_service.dart
import 'package:flutter/foundation.dart';
import '../models/user.dart';

/// Mock authentication service
/// In production, this will integrate with Firebase Auth or Auth0 with 2FA
class AuthService extends ChangeNotifier {
  User? _currentUser;
  bool _isAuthenticated = false;

  User? get currentUser => _currentUser;
  bool get isAuthenticated => _isAuthenticated;

  /// Mock login with 2FA simulation
  /// In production: integrate Firebase Auth with TOTP/SMS 2FA
  Future<bool> login(String email, String password) async {
    // Simulate API delay
    await Future.delayed(const Duration(seconds: 1));

    // Mock validation (in production, use Firebase/Auth0)
    if (email.isNotEmpty && password.length >= 6) {
      _currentUser = User(
        id: 'mock-user-id',
        email: email,
        fullName: 'Demo User',
        twoFactorEnabled: true,
        createdAt: DateTime.now(),
      );
      _isAuthenticated = true;
      notifyListeners();
      return true;
    }
    return false;
  }

  /// Mock 2FA verification
  /// In production: verify TOTP code or SMS code
  Future<bool> verify2FA(String code) async {
    await Future.delayed(const Duration(seconds: 1));
    
    // Mock: accept any 6-digit code
    if (code.length == 6) {
      return true;
    }
    return false;
  }

  /// Mock signup
  Future<bool> signup(String email, String password, String fullName) async {
    await Future.delayed(const Duration(seconds: 1));

    if (email.isNotEmpty && password.length >= 6 && fullName.isNotEmpty) {
      _currentUser = User(
        id: 'mock-user-id-${DateTime.now().millisecondsSinceEpoch}',
        email: email,
        fullName: fullName,
        twoFactorEnabled: false,
        createdAt: DateTime.now(),
      );
      _isAuthenticated = true;
      notifyListeners();
      return true;
    }
    return false;
  }

  /// Logout
  Future<void> logout() async {
    _currentUser = null;
    _isAuthenticated = false;
    notifyListeners();
  }

  /// Enable 2FA (mock)
  Future<String> enable2FA() async {
    await Future.delayed(const Duration(seconds: 1));
    // In production: generate TOTP secret and return QR code URL
    return 'otpauth://totp/SmartPlugAI:${_currentUser?.email}?secret=MOCKBASE32SECRET&issuer=SmartPlugAI';
  }

  /// Disable 2FA (mock)
  Future<bool> disable2FA(String password) async {
    await Future.delayed(const Duration(seconds: 1));
    if (password.isNotEmpty) {
      return true;
    }
    return false;
  }
}
